explainaboard/metrics/nlg_meta_evaluation.py

```diff
@@ -18,6 +19,142 @@
 from explainaboard.utils.typing_utils import narrow, unwrap_or
 
 
+@dataclass
+@common_registry.register("CorrelationNLGConfig")
+class CorrelationNLGConfig(MetricConfig):
+    """Configuration of a correlation for general NLG tasks.
+
+    Args:
+        level: there are following levels: sample level, system level and dataset level
+                See more details: https://aclanthology.org/2020.emnlp-main.751.pdf
+        func_name: the method for calculating the correlation (e.g., Spearman)
+    """
+
+    level: str = "sample"
+    func_name: str = "spearmanr"
+
+    def to_metric(self) -> Metric:
+        """See MetricConfig.to_metric."""
+        return CorrelationNLG(self)
+
+    def get_correlation_func(self, name: str):
+        """Get correlation function based on function name.
+
+        TODO(pengfei): organize this in a better way
+
+        Args:
+            name: function name
+        """
+        if name == "spearmanr":
+            return stats.spearmanr
+        elif name == "pearsonr":
+            return stats.pearsonr
+        elif name == "kendalltau":
+            return stats.kendalltau
+        else:
+            raise ValueError(f"The correlation function {name} hasn't been supported")
+
+
+class CorrelationNLG(Metric):
+    """A metric that calculates correlations."""
+
+    n_samples = 0
+    n_systems = 0
+
+    def is_simple_average(self, stats: MetricStats) -> bool:
+        """See Metric.is_simple_average."""
+        return False
+
+    def uses_customized_aggregate(self) -> bool:
+        """See Metric.uses_customized_aggregate."""
+        return True
+
+    def calc_stats_from_data(
+        self,
+        true_data: list[list[float]],
+        pred_data: list[list[float]],
+    ) -> MetricStats:
+        """See Metric.calc_stats_from_data."""
+        config = narrow(CorrelationNLGConfig, self.config)
+        self.n_samples = len(true_data)
+        self.n_systems = len(true_data[0])
+
+        if config.level == "sample":
+            corr_func = config.get_correlation_func(config.func_name)
+            return SimpleMetricStats(
+                np.array(
+                    [
+                        corr_func(true, pred)[0]
+                        for true, pred in zip(true_data, pred_data)
+                    ]
+                )
+            )
+        elif config.level == "system":
+            return SimpleMetricStats(
+                np.array([true + pred for true, pred in zip(true_data, pred_data)])
+            )
+        else:  # TODO(Pengfei): better way to organize different levels
+            return SimpleMetricStats(
+                np.array(
+                    [[true[0], pred[0]] for true, pred in zip(true_data, pred_data)]
+                )
+            )
```

Comment:
As far as I read this class, there are only dataset/system/sample choices for `level`. The class has to implement a full list of conditions for these values, and raise errors for any other values.